### PR TITLE
Generate reference from snapcraft's "help" command.

### PR DESCRIPTION
**Pull request by dholbach**
Run:
```
  ./snapcraft/scripts/generate_reference.py
```
or
```
  ./snapcraft/scripts/generate_reference.py --output-file=/tmp/something.md
```
LP: #1544540

### CODE REVIEW

**Comment by dholbach** [comment]
This is just a first cut to get an idea of you generally like the idea. Having submitted this I wonder if I shouldn't generate just one docs/reference.md file with an auto-generated ToC instead...

**Comment by dholbach** [comment]
Ah... maybe I should explain the rationale for this: it'd be great if (once the md importer for developer.u.c finally lands) we could show an automatically updated reference without having to duplicate this information.

**Comment by dholbach** [comment]
I changed it to only write to one file.

Let me know if you want me to break out the functionality into a separate file outside `setup.py`.

File: setup.py [hunk 1]
```diff
@@ -31,6 +33,99 @@
...
+        for topic in self.topics:
+            text = _get_help_output(topic)
+            if not text.startswith('The plugin has no documentation'):
+                self.docs += [
+                    (topic, text)
+                ]
+        self._write_index_doc()
+
+    def _write_index_doc(self):
+        text = '''
```
Comment (sergiusens) [inline setup.py]: Use three """ instead, also, maybe call this text_template or something... @elopio will have ideas wrt to consistency on this one.

File: setup.py [hunk 2]
```diff
@@ -31,6 +33,99 @@
         version = match.group(1)
 
 
+def _get_help_output(topic, lines=False):
+    TRANSLATE = {
+        'tar-content': 'tar_content',
+    }
```
Comment (elopio) [inline setup.py]: Does this apply to all the topics with a dash? If so, you could just replace the dash.

File: setup.py [hunk 3]
```diff
@@ -31,6 +33,99 @@
...
+    if topic in TRANSLATE:
+        topic = TRANSLATE[topic]
+    return _run_snapcraft(['help', topic], lines)
+
+
+def _run_snapcraft(cmd, lines=False):
+    cmd = ['./bin/snapcraft']+cmd
```
Comment (elopio) [inline setup.py]: pep8 doesn't check this file, which is probably a bug. You should use spaces between the operators:
cmd = ['./bin/snapcraft'] + cmd

File: setup.py [hunk 4]
```diff
@@ -31,6 +33,99 @@
...
+    command = subprocess.Popen(cmd, stdout=subprocess.PIPE)
+    output = command.communicate()[0]
```
Comment (elopio) [inline setup.py]: why don't you use subprocess.check_output here?

File: setup.py [hunk 5]
```diff
@@ -31,6 +33,99 @@
...
+
+    if sys.version_info.major == 3:
+        output = output.decode('utf-8')  # bytes to str
```
Comment (elopio) [inline setup.py]: I think I would prefer just to ignore python2. @sergiusens do you know if there's a good reason to keep this check?

File: setup.py [hunk 6]
```diff
@@ -31,6 +33,99 @@
...
+    def initialize_options(self):
+        self.output_file = None
+
+    def finalize_options(self):
+        if not self.output_file:
+            self.output_file = os.path.abspath(self.DEFAULT_REFERENCE_FILE)
+
+    def run(self):
+        assert os.getcwd() == os.path.abspath(os.path.dirname(__file__)), \
+            'Must be run in snapcraft root.'
```
Comment (elopio) [inline setup.py]: This might be confusing. Root is also the user. So maybe:
'Must be run in the root directory of snapcraft.'

**Comment by elopio** [comment]
I like the idea. I don't like that it's in the setup.py.
Could we move it to something like snapcraft/scripts/generate_reference.py ?

File: setup.py (hunk 2 above)
Comment (dholbach) [inline setup.py]: Right now it's the only one and I thought I'd rather make this explicit than just changing anything which comes up with a dash.

File: setup.py (hunk 5 above)
Comment (dholbach) [inline setup.py]: We might run this on developer.u.c where I don't know which python versions are available.

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by dholbach** [comment]
I hope it's all fine now.

**Comment by elopio** [comment]
Looks good. Your replies make a lot of sense, and thanks for moving it.
Now that you are inside the snapcraft, the static tests will check the script. (it was a trap! :)
Travis is giving this error:
snapcraft/scripts/generate_reference.py:38:9: E128 continuation line under-indented for visual indent

**Comment by dholbach** [comment]
It's fixed now.

File: setup.py (hunk 5 above)
Comment (dholbach) [inline setup.py]: I removed the python2-ism. We have python3 available.

**Comment by elopio** [comment]
add to whitelist

**Comment by dholbach** [comment]
What does "add to whitelist" mean? Do I need to do anything?

**Comment by kyrofa** [comment]
@dholbach I *think* he's experimenting with the bot.

**Comment by elopio** [comment]
retest this please

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by elopio** [comment]
landing...

**Pull request merged**